include/exadg/operators/variable_coefficients.h

```diff
@@ -23,168 +23,231 @@
+   * according to the desired properties @p use_face_loops_in, @p use_neighbor_faces_in,
+   * and @p use_cell_based_face_loops_in, i.e. resizes the desired tables to the required number of
+   * cells, faces, and quadrature points.
+   *
+   * @param matrix_free Underlying matrix-free description
+   * @param quad_index Quadrature index in the matrix-free description to create the
+   * coefficient tables according to the quadrature points
+   * @param use_face_loops_in Boolean switch to use a coefficient table for faces
+   * @param use_neighbor_faces_in Boolean switch to use a coefficient table for neighbors, in case
+   * the coefficients can differ on the two sides of a face
+   * @param use_cell_based_face_loops_in Boolean switch to use a coefficient table for cell-based
+   * face access
+   */
   template<int dim, typename Number>
   void
   initialize(dealii::MatrixFree<dim, Number> const & matrix_free,
              unsigned int const                      quad_index,
-             coefficient_type const &                constant_coefficient)
-  {
-    reinit(matrix_free, quad_index);
-
-    fill(constant_coefficient);
-  }
-
-  coefficient_type
-  get_coefficient(unsigned int const cell, unsigned int const q) const
-  {
-    return coefficients_cell[cell][q];
-  }
-
-  void
-  set_coefficient(unsigned int const cell, unsigned int const q, coefficient_type const & value)
-  {
-    coefficients_cell[cell][q] = value;
-  }
-
-private:
-  template<int dim, typename Number>
-  void
-  reinit(dealii::MatrixFree<dim, Number> const & matrix_free, unsigned int const quad_index)
+             bool const                              use_face_loops_in,
+             bool const                              use_neighbor_faces_in,
+             bool const                              use_cell_based_face_loops_in)
   {
-    coefficients_cell.reinit(matrix_free.n_cell_batches(), matrix_free.get_n_q_points(quad_index));
-  }
+    use_face_loops            = use_face_loops_in;
+    use_neighbor_faces        = use_neighbor_faces_in;
+    use_cell_based_face_loops = use_cell_based_face_loops_in;
 
-  void
-  fill(coefficient_type const & constant_coefficient)
-  {
-    coefficients_cell.fill(constant_coefficient);
+    reinit(matrix_free, quad_index);
   }
 
-  // variable coefficients
-  dealii::Table<2, coefficient_type> coefficients_cell;
-};
-
-template<typename coefficient_type>
-class VariableCoefficients
-{
-public:
+  /**
+   * Same as above, but fills the coefficient tables with the given @p constant_coefficient.
+   */
   template<int dim, typename Number>
   void
   initialize(dealii::MatrixFree<dim, Number> const & matrix_free,
              unsigned int const                      quad_index,
+             bool const                              use_face_loops_in,
+             bool const                              use_neighbor_faces_in,
+             bool const                              use_cell_based_face_loops_in,
```

Comment:
I am not sure if duplicating the above interface here is desirable. Shouldn't we rather have only one initialize function and then call a second function in user code to set/fill the coefficient tables?

Reply:
Both make sense to me. I can happily make it more explicit with a separate function.